File: 07-selecao-enumeracoes-e-estruturas/exemplos/semaforo.cpp
#include "bscpp.hpp"

// Análise
//
// Calcular a próxima cor de uma semaforo que tem as cores verde, amarelo e vermelho.
//
// Definição de tipos de dados

// Representa a cor de um semáforo.
enum Cor {
    Verde,
    Vermelho,
    Amarelo,
};


// Produz a próximo cor de um semáforo que está com a cor c.
Cor proxima_cor(Cor c) {
    Cor proxima;
    switch (c) {
        case Verde:
            proxima = Amarelo;
            break;
        case Vermelho:
            proxima = Verde;
            break;
        case Amarelo:
            proxima = Vermelho;
            break;
    }
    return proxima;
}

examples
{
    check_expect(proxima_cor(Verde), Amarelo);
    check_expect(proxima_cor(Vermelho), Verde);
    check_expect(proxima_cor(Amarelo), Vermelho);
}


int main()
{
    run_tests();
}

File: 05-tipos-de-dados/exemplos/semaforo.py
from enum import Enum, auto


class Cor(Enum):
    '''O cor de um semáforo de trânsito.'''
    VERDE = auto()
    VERMELHO = auto()
    AMARELO = auto()


def proxima_cor(atual: Cor) -> Cor:
    '''
    Produz a pŕoxima cor de uma semáfaro que está na cor *atual*.

    Exemplos
    >>> proxima_cor(Cor.VERDE).name
    'AMARELO'
    >>> proxima_cor(Cor.AMARELO).name
    'VERMELHO'
    >>> proxima_cor(Cor.VERMELHO).name
    'VERDE'
    '''
    if atual == Cor.VERDE:
        proxima = Cor.AMARELO
    elif atual == Cor.AMARELO:
        proxima = Cor.VERMELHO
    elif atual == Cor.VERMELHO:
        proxima = Cor.VERDE
    return proxima
